packages/buidler-core/src/internal/buidler-evm/jsonrpc/batch.ts

```diff
@@ -0,0 +1,75 @@
+import { JsonRpcRequest } from "../../util/jsonrpc";
+
+import { HttpRequestService } from "./http";
+
+type BatchJsonRpcRequest = JsonRpcRequest[];
+type ResolveFunction = (value?: unknown) => void;
+type RejectFunction = (error?: any) => void;
+
+interface DeferredRequest {
+  body: JsonRpcRequest;
+  resolve: RejectFunction;
+  reject: RejectFunction;
+}
+
+export interface JsonRpcSender {
+  send(method: string, params?: any[]): Promise<any>;
+}
+
+export class JsonRpcRequestBatcher implements JsonRpcSender {
+  private _deferredRequests: DeferredRequest[] = [];
+  private _nextRequestId = 1;
+
+  constructor(
+    private readonly _httpService: HttpRequestService,
+    private readonly _batchingTime: number = 0
+  ) {}
+
+  public send(method: string, params?: any[]): Promise<any> {
+    const request = this._getJsonRpcRequest(method, params);
+    return this._deferSend(request);
+  }
+
+  private _deferSend(request: JsonRpcRequest) {
+    return new Promise((resolve, reject) => {
+      this._deferredRequests.push({ body: request, resolve, reject });
+
+      if (this._deferredRequests.length === 1) {
+        setTimeout(this._performSend.bind(this), this._batchingTime);
+      }
+    });
+  }
+
+  private async _performSend() {
+    const requests = this._deferredRequests.map((req) => req.body);
+    const resolves = this._deferredRequests.map((req) => req.resolve);
+    const rejects = this._deferredRequests.map((req) => req.reject);
+    this._deferredRequests = [];
+    try {
+      const responses = await this._httpService.send(requests);
+      for (let i = 0; i < responses.length; i++) {
+        if (responses[i] instanceof Error) {
+          rejects[i](responses[i]);
+        } else {
+          resolves[i](responses[i]);
+        }
+      }
+    } catch (e) {
+      for (const reject of rejects) {
```

Comment:
This is a bit dangerous, as you can easily end up in a situation where a promise's `reject` is called twice, or its `reject` and `resolve` are called.

Reply:
You are right. I'll investigate a better way to handle this